internal/commands/build.go

```diff
@@ -302,3 +320,81 @@ func (c *OneBuildConfig) New() LoadableConfig {
 func OneBuildConfigFrom(cmd *cli.Command) (*OneBuildConfig, error) {
 	return ConfigFromCLI[*OneBuildConfig](oneBuildConfigKey, cmd)
 }
+
+func uploadToUrl(uploadUrl string, uploadBodyParams []shared.UploadBodyParams, filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+	fileSize := fileInfo.Size()
+
+	var requestBody bytes.Buffer
+	multipartWriter := multipart.NewWriter(&requestBody)
+
+	for _, param := range uploadBodyParams {
+		_ = multipartWriter.WriteField(param.Key, param.Value)
+	}
+
+	fileWriter, err := multipartWriter.CreateFormFile("file", fileInfo.Name())
+	if err != nil {
+		return err
+	}
+
+	progressReader := &progressReader{
+		reader: file,
+		total:  fileSize,
+		callback: func(percentage float64, loaded int64, total int64) {
+			fmt.Printf("Upload progress: %.2f%% (%d/%d bytes)\r", percentage, loaded, total)
+		},
+	}
+
+	_, err = io.Copy(fileWriter, progressReader)
+	if err != nil {
+		return err
+	}
+	print("\n")
+
+	multipartWriter.Close()
+
+	req, err := http.NewRequest("POST", uploadUrl, &requestBody)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", multipartWriter.FormDataContentType())
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("upload failed with status: %s", resp.Status)
+	}
+
+	return nil
+}
+
+type progressReader struct {
+	reader   io.Reader
+	total    int64
+	read     int64
+	callback func(percentage float64, loaded int64, total int64)
+}
+
+func (pr *progressReader) Read(p []byte) (int, error) {
+	n, err := pr.reader.Read(p)
```

Comment:
By default `io.Copy` chunks with a 4096-byte buffer, so this might noisy.

Reply:
i think it's nice for it to be noisy--each progress logline overwrites itself so it still looks nice in the terminal